content/v2.0/actions/parameters.md

```diff
@@ -0,0 +1,344 @@
+Please check the related documentation for a complete list of options and how to share code between validations.
+
+## Concrete Classes
+
+The params DSL is really quick and intuitive but it has the drawback that it can be visually noisy and makes it hard to unit test.
+An alternative is to extract a class and pass it as an argument to `.params`.
+
+```ruby
+# app/actions/signup/my_params.rb
+
+module Web
+  module Controllers
+    module Signup
+      class MyParams < Web::Action::Params
+        MEGABYTE = 1024 ** 2
+
+        params do
+          required(:name).filled(:str?)
+          required(:email).filled(:str?, format?: /@/)
+          required(:password).filled(:str?)
+          required(:terms_of_service).filled(:bool?)
+          required(:age).filled(:int?, included_in?: 18..99)
+          optional(:avatar).filled(size?: 1..(MEGABYTE * 3)
+        end
+      end
+    end
+  end
+end
+```
+
+```ruby
+# apps/web/controllers/signup/create.rb
+require_relative './my_params'
+
+module Sandbox
+  module Actions
+    module Signup
+      class Create < Action
+        params MyParams
+
+        def handle(req, res)
+          if req.params.valid?
+            # ...
+          else
+            # ...
+          end
+        end
+      end
+    end
+  end
+end
+```
+
+### Inline predicates
+
+In case there is a predicate that is needed only for the current params, you can define inline predicates:
+
+```ruby
+# app/actions/books/create.rb
+
+module Sandbox
+  module Actions
+    module Books
+      class Create < Action
+        params Class.new(Hanami::Action::Params) {
+          predicate(:cool?, message: "is not cool") do |current|
+            current.match(/cool/)
+          end
+
+          validations do
+            required(:book).schema do
+              required(:title) { filled? & str? & cool? }
+            end
+          end
+        }
+
+        def handle(req, res)
+          if req.params.valid?
+            res.body = 'OK'
+          else
+            res.body = params.error_messages.join("\n")
+          end
+        end
+      end
+    end
+  end
+end
+```
+
+
+
+## Body Parsers
+
+Rack ignores request bodies unless they come from a form submission.
+If we have a JSON endpoint, the payload isn't available in `params`.
+
+```ruby
+# apps/web/controllers/books/create.rb
+
+module Sandbox
+  module Actions
+    module Books
+      class Create < Action
+        accept :json
+
+        def handle(req, res)
+          res.body = req.params.to_h # => {}
+        end
+      end
+    end
+  end
+end
+```
+
+```shell
+$ curl http://localhost:2300/books      \
+    -H "Content-Type: application/json" \
+    -H "Accept: application/json"       \
+    -d '{"book":{"title":"Hanami"}}'    \
+    -X POST
+```
+
+To properly coerce `json` params, you should enable this feature.
+
+```ruby
+#config.ru
+require "hanami/middleware/body_parser"
```

Comment:
@timriley Here is a problem: `hanami/middleware/body_parser` requires a `hanami/middleware/error` - but it's not automatically loaded, and when I include this line, it returns an error.

```shell
/Users/sebastian/.asdf/installs/ruby/3.0.2/lib/ruby/gems/3.0.0/gems/hanami-router-2.0.0.beta1/lib/hanami/middleware/body_parser/errors.rb:12:in `<class:BodyParser>': uninitialized constant Hanami::Middleware::Error (NameError)
Did you mean?  Hanami::Error
               IOError
               Errno
```

Reply:
It's already reported in: https://trello.com/c/dh4Wtjea/110-body-parsing-should-be-built-into-hanami-controller-not-be-solely-a-standalone-middleware

So resolving this for now.